plugins/anomaly-detection/AnomalyDetection.py

```diff
@@ -0,0 +1,615 @@
+
+    # Schema for the Dataframe to ve created from the RDD
+    rdd3Schema = StructType([
+        StructField('sourceIP', StringType(), True),
+        StructField('sourceTransportPort', LongType(), True),
+        StructField('destinationIP', StringType(), True),
+        StructField('destinationTransportPort', LongType(), True),
+        StructField('protocolIdentifier', LongType(), True),
+        StructField('flowStartSeconds', TimestampType(), True),
+        StructField('flowEndSeconds', ArrayType(TimestampType(), True)),
+        StructField('throughputStandardDeviation', DoubleType(), True),
+        StructField('algoType', StringType(), True),
+        StructField('algoCalc', ArrayType(DoubleType(), True)),
+        StructField('Throughputs', ArrayType(DecimalType(38, 18), True)),
+        StructField('anomaly', ArrayType(BooleanType(), True))
+    ])
+
+    plotDF5 = spark.createDataFrame(rdd3, rdd3Schema)
+    ret_plotDF = filter_df_with_true_anomalies(spark, plotDF5, algo_type)
+    # Write anomalous records to DB/CSV - Module WIP
+    # Module to write to CSV. Optional.
+    ret_plotDF = ret_plotDF.withColumn(
+        "flowEndSeconds",
+        f.col("flowEndSeconds").cast("string")).withColumn(
+        "throughputStandardDeviation",
+        f.col("throughputStandardDeviation").cast("string")).withColumn(
+        "algoCalc", f.col("algoCalc").cast("string")).withColumn(
+        "Throughputs", f.col("Throughputs").cast("string")).withColumn(
+        "anomaly", f.col("anomaly").cast(
+            "string"))
+    ret_plotDF = ret_plotDF.withColumn('id', f.lit(str(tad_id_input)))
+    return ret_plotDF
+
+
+def anomaly_detection(algo_type, db_jdbc_address, start_time, end_time,
+                      tad_id_input):
+    spark = SparkSession.builder.getOrCreate()
+    sql_query = ("select {} from {} "
+                 .format(", ".join(FLOW_TABLE_COLUMNS), table_name))
+    if start_time:
+        sql_query += "WHERE flowStartSeconds >= '{}' ".format(start_time)
+    if end_time:
+        if start_time:
+            sql_query += "AND flowEndSeconds < '{}' ".format(end_time)
+        else:
+            sql_query += "WHERE flowEndSeconds < '{}' ".format(end_time)
+    sql_query += "group by {} ".format(", ".join(GROUP_COLUMNS))
+
+    plotDF = (
+        spark.read.format("jdbc").option(
+            'driver', "ru.yandex.clickhouse.ClickHouseDriver").option(
+            "url", db_jdbc_address).option(
+            "user", os.getenv("CH_USERNAME")).option(
+            "password", os.getenv("CH_PASSWORD")).option(
+            "query", sql_query).load()
+    )
+
+    plotDF1 = plotDF.groupby("sourceIP", "sourceTransportPort",
+                             "destinationIP", "destinationTransportPort",
+                             "protocolIdentifier", "flowStartSeconds").agg(
+        f.stddev_samp("max(throughput)").alias(
+            "throughputStandardDeviation"))
+    plotDF2 = plotDF.groupby("sourceIP", "sourceTransportPort",
```

Comment:
plotDF.groupby() has been done multiple times, please create another DF to cache result and reuse later.

Reply:
ack